### ISSUE TITLE
softhsm does not enforce lockout?

### ISSUE THREAD

**Issue by nimbius** [opendnssec/SoftHSMv2#399]
Unlike traditional HSM's, softhsm2 doesnt appear to enforce a lockout.  I can bang away with 
` pkcs11-tool --module /usr/lib/softhsm/libsofthsm2.so -l -t -O`
and after numerous incorrect entries can still access the slot and perform an OK authentication against it.

**Comment by bellgrim** [comment]
No, there is no limit on how many failed login attempts you can do. The reason is that anyone who can access SoftHSM and log in, can also extract the password hash from the token file and do offline attacks without having to worry about the lockout mechanism.

A lockout mechanism would be helpful if there were a separation between the user and the token file, e.g. a middleware with a remote connection or interprocess communication. That middleware could, however, implement a lockout mechanism on top of SoftHSM.

**Comment by nimbius** [comment]
Understood.  If this is the case, then is there a purpose for the SO pin other than for the user to achieve familiarity with the concept as it applies to a real HSM?  Im having difficulty seeing how softHSM provides meaningful security if it stored, for example, a CA certificate used to sign CSRs.

Thank you for your help

**Comment by bellgrim** [comment]
We use the SO since it is required by PKCS#11, e.g. to initialize a token.

SoftHSM will not provide the same level of security as a real HSM. Using SoftHSM is similar to using an encrypted private key file for your CA.
